File: sdk/squaremind-sdk/src/agent.ts
import { EventEmitter } from 'events';
import {
  AgentConfig,
  AgentState,
  Capability,
  CapabilityType,
  Reputation,
  SquaremindIdentity,
  Task,
  TaskResult,
  TaskStatus,
} from './types';

/**
 * Generate a UUID v4
 */
function generateUUID(): string {
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
    const r = (Math.random() * 16) | 0;
    const v = c === 'x' ? r : (r & 0x3) | 0x8;
    return v.toString(16);
  });
}

/**
 * Squaremind Agent - An autonomous AI agent
 */
export class Agent extends EventEmitter {
  public readonly identity: SquaremindIdentity;
  public readonly capabilities: Map<CapabilityType, Capability>;
  public reputation: Reputation;
  public state: AgentState;
  public currentTask: Task | null = null;

  private model: string;
  private taskQueue: Task[] = [];

  constructor(config: AgentConfig) {
    super();

    // Create identity
    this.identity = {
      sid: generateUUID(),
      name: config.name,
      publicKey: this.generatePublicKey(),
      createdAt: new Date(),
      parentSid: config.parentSid,
      generation: config.parentSid ? 1 : 0,
    };

    // Initialize capabilities
    this.capabilities = new Map();
    for (const capType of config.capabilities) {
      this.capabilities.set(capType, {
        type: capType,
        proficiency: 0.5, // Start at 50%
      });
    }

    // Initialize reputation
    this.reputation = {
      overall: 50,
      reliability: 50,
      quality: 50,
      cooperation: 50,
      honesty: 50,
      tasksCompleted: 0,
      tasksFailed: 0,
      lastActive: new Date(),
    };

    this.state = 'initializing';
    this.model = config.model || 'claude-sonnet-4-20250514';
  }

  private generatePublicKey(): string {
    // Simplified - in production, use proper Ed25519
    const bytes = new Uint8Array(32);
    if (typeof crypto !== 'undefined' && crypto.getRandomValues) {
      crypto.getRandomValues(bytes);
    } else {
      for (let i = 0; i < 32; i++) {
        bytes[i] = Math.floor(Math.random() * 256);
      }
    }
    return Array.from(bytes)
      .map((b) => b.toString(16).padStart(2, '0'))
      .join('');
  }

  /**
   * Start the agent
   */
  async start(): Promise<void> {
    this.state = 'idle';
    this.emit('started');
    this.processQueue();
  }

  /**
   * Stop the agent
   */
  stop(): void {
    this.state = 'terminated';
    this.emit('stopped');
  }

  /**
   * Pause the agent
   */
  pause(): void {
    if (this.state === 'idle') {
      this.state = 'paused';
      this.emit('paused');
    }
  }

  /**
   * Resume the agent
   */
  resume(): void {
    if (this.state === 'paused') {
      this.state = 'idle';
      this.emit('resumed');
      this.processQueue();
    }
  }

  /**
   * Submit a task to the agent
   */
  submitTask(task: Task): void {
    this.taskQueue.push(task);
    this.emit('task:received', task);

    if (this.state === 'idle') {
      this.processQueue();
    }
  }

  /**
   * Get capability match score
   */
  getCapabilityScore(required: CapabilityType[]): number {
    if (required.length === 0) return 1.0;

    let totalScore = 0;
    let matched = 0;

    for (const req of required) {
      const cap = this.capabilities.get(req);
      if (cap) {
        totalScore += cap.proficiency;
        matched++;
      }
    }

    if (matched === 0) return 0;

    const coverage = matched / required.length;
    const avgProficiency = totalScore / matched;

    return coverage * avgProficiency;
  }

  /**
   * Check if agent has a capability
   */
  hasCapability(capType: CapabilityType): boolean {
    return this.capabilities.has(capType);
  }

  private async processQueue(): Promise<void> {
    while (this.taskQueue.length > 0 && this.state === 'idle') {
      const task = this.taskQueue.shift();
      if (task) {
        await this.executeTask(task);
      }
    }
  }

  private async executeTask(task: Task): Promise<TaskResult> {
    this.state = 'working';
    this.currentTask = task;
    this.reputation.lastActive = new Date();

    const startTime = Date.now();

    try {
      // Simulate task execution
      // In real implementation, this would call the LLM
      const output = await this.performTask(task);

      const result: TaskResult = {
        taskId: task.id,
        agentSid: this.identity.sid,
        status: 'completed',
        output,
        quality: 0.8,
        duration: Date.now() - startTime,
        timestamp: new Date(),
      };

      this.recordSuccess(result.quality);
      this.emit('task:completed', result);

      this.state = 'idle';
      this.currentTask = null;

      return result;
    } catch (error) {
      const result: TaskResult = {
        taskId: task.id,
        agentSid: this.identity.sid,
        status: 'failed',
        output: '',
        error: error instanceof Error ? error.message : 'Unknown error',
        quality: 0,
        duration: Date.now() - startTime,
        timestamp: new Date(),
      };

      this.recordFailure();
      this.emit('task:failed', result);

      this.state = 'idle';
      this.currentTask = null;

      return result;
    }
  }

  private async performTask(task: Task): Promise<string> {
    // Simulated task execution
    // In real implementation, this would use the LLM provider
    return `[Simulated] Completed: ${task.description}`;
  }

  private recordSuccess(quality: number): void {
    this.reputation.tasksCompleted++;
    this.reputation.quality =
      this.reputation.quality * 0.9 + quality * 100 * 0.1;
    this.reputation.reliability =
      this.reputation.reliability * 0.95 + 100 * 0.05;
    this.recalculateOverall();
  }

  private recordFailure(): void {
    this.reputation.tasksFailed++;
    this.reputation.reliability *= 0.9;
    this.recalculateOverall();
  }

  private recalculateOverall(): void {
    this.reputation.overall =
      (this.reputation.reliability +
        this.reputation.quality +
        this.reputation.cooperation +
        this.reputation.honesty) /
      4;
  }
}

File: pkg/agent/agent.go
package agent

import (
	"context"
	"fmt"
	"sync"
	"time"

	"github.com/square-mind/squaremind/pkg/identity"
	"github.com/square-mind/squaremind/pkg/llm"
)

// AgentState represents the current state of an agent
type AgentState string

const (
	StateInitializing AgentState = "initializing"
	StateIdle         AgentState = "idle"
	StateWorking      AgentState = "working"
	StatePaused       AgentState = "paused"
	StateTerminated   AgentState = "terminated"
)

// Agent represents a squaremind AI agent
type Agent struct {
	mu sync.RWMutex

	// Identity
	Identity *identity.SquaremindIdentity

	// Capabilities
	Capabilities *identity.CapabilitySet

	// LLM Backend
	Provider llm.Provider
	Model    string

	// State
	State       AgentState
	CurrentTask *Task

	// Reputation
	Reputation *Reputation

	// Memory
	Memory *AgentMemory

	// Channels for coordination
	taskChan   chan *Task
	resultChan chan *TaskResult
	stopChan   chan struct{}

	// Lifecycle
	StartedAt  time.Time
	LastActive time.Time
}

// AgentConfig holds configuration for creating a new agent
type AgentConfig struct {
	Name         string
	Capabilities []identity.CapabilityType
	Provider     llm.Provider
	Model        string
	ParentSID    string
}

// NewAgent creates a new squaremind agent
func NewAgent(cfg AgentConfig) (*Agent, error) {
	// Create identity
	id, err := identity.NewSquaremindIdentity(cfg.Name, cfg.ParentSID)
	if err != nil {
		return nil, err
	}

	// Initialize capabilities
	capSet := identity.NewCapabilitySet()
	for _, capType := range cfg.Capabilities {
		capSet.Add(&identity.Capability{
			Type:        capType,
			Proficiency: 0.5, // Start at 50%, improve through tasks
		})
	}

	return &Agent{
		Identity:     id,
		Capabilities: capSet,
		Provider:     cfg.Provider,
		Model:        cfg.Model,
		State:        StateInitializing,
		Reputation:   NewReputation(),
		Memory:       NewAgentMemory(),
		taskChan:     make(chan *Task, 10),
		resultChan:   make(chan *TaskResult, 10),
		stopChan:     make(chan struct{}),
		StartedAt:    time.Now(),
		LastActive:   time.Now(),
	}, nil
}

// Start begins the agent's autonomous operation
func (a *Agent) Start(ctx context.Context) error {
	a.mu.Lock()
	a.State = StateIdle
	a.mu.Unlock()

	go a.runLoop(ctx)
	return nil
}

// runLoop is the main agent operation loop
func (a *Agent) runLoop(ctx context.Context) {
	for {
		select {
		case <-ctx.Done():
			a.terminate()
			return
		case <-a.stopChan:
			a.terminate()
			return
		case task := <-a.taskChan:
			a.executeTask(ctx, task)
		}
	}
}

// executeTask handles task execution
func (a *Agent) executeTask(ctx context.Context, task *Task) {
	a.mu.Lock()
	a.State = StateWorking
	a.CurrentTask = task
	a.LastActive = time.Now()
	a.mu.Unlock()

	startTime := time.Now()

	// Execute with LLM
	result, err := a.performTask(ctx, task)
	result.Duration = time.Since(startTime)
	result.Timestamp = time.Now()
	result.AgentSID = a.Identity.SID

	a.mu.Lock()
	a.State = StateIdle
	a.CurrentTask = nil
	a.mu.Unlock()

	// Update reputation based on result
	if err != nil {
		a.Reputation.RecordFailure()
	} else {
		a.Reputation.RecordSuccess(result.Quality)
	}

	// Add to episodic memory
	a.Memory.AddEpisode(Episode{
		Type:    "task_completion",
		Content: fmt.Sprintf("Completed task: %s", task.Description),
		Context: map[string]interface{}{
			"task_id": task.ID,
			"quality": result.Quality,
			"status":  result.Status,
		},
		Salience: result.Quality,
	})

	// Send result
	select {
	case a.resultChan <- result:
	default:
		// Channel full, drop result
	}
}

// performTask uses the LLM to perform the actual task
func (a *Agent) performTask(ctx context.Context, task *Task) (*TaskResult, error) {
	// If no provider, return simulated result
	if a.Provider == nil {
		return &TaskResult{
			TaskID:  task.ID,
			Status:  TaskCompleted,
			Output:  fmt.Sprintf("[Simulated] Completed: %s", task.Description),
			Quality: 0.75,
		}, nil
	}

	prompt := a.buildPrompt(task)

	response, err := a.Provider.Complete(ctx, llm.CompletionRequest{
		Model:  a.Model,
		Prompt: prompt,
	})
	if err != nil {
		return &TaskResult{
			TaskID: task.ID,
			Status: TaskFailed,
			Error:  err.Error(),
		}, err
	}

	return &TaskResult{
		TaskID:  task.ID,
		Status:  TaskCompleted,
		Output:  response.Content,
		Quality: 0.8, // Would be evaluated by quality assessment
	}, nil
}

// buildPrompt constructs the prompt for the LLM
func (a *Agent) buildPrompt(task *Task) string {
	capsJSON := a.Capabilities.ToJSON()

	return fmt.Sprintf(`You are a squaremind AI agent with the following identity:
Name: %s
SID: %s
Capabilities: %s

Your task:
%s

Requirements:
%s

Perform this task to the best of your ability. Be thorough and precise.`,
		a.Identity.Name,
		a.Identity.SID,
		capsJSON,
		task.Description,
		task.Requirements,
	)
}

// Stop signals the agent to stop
func (a *Agent) Stop() {
	close(a.stopChan)
}

// terminate cleans up agent resources
func (a *Agent) terminate() {
	a.mu.Lock()
	defer a.mu.Unlock()
	a.State = StateTerminated
}

// SubmitTask submits a task to the agent
func (a *Agent) SubmitTask(task *Task) {
	select {
	case a.taskChan <- task:
	default:
		// Channel full
	}
}

// GetResults returns the results channel
func (a *Agent) GetResults() <-chan *TaskResult {
	return a.resultChan
}

// GetState returns the current agent state
func (a *Agent) GetState() AgentState {
	a.mu.RLock()
	defer a.mu.RUnlock()
	return a.State
}

// GetCurrentTask returns the current task being worked on
func (a *Agent) GetCurrentTask() *Task {
	a.mu.RLock()
	defer a.mu.RUnlock()
	return a.CurrentTask
}

// Pause pauses the agent
func (a *Agent) Pause() {
	a.mu.Lock()
	defer a.mu.Unlock()
	if a.State == StateIdle {
		a.State = StatePaused
	}
}

// Resume resumes a paused agent
func (a *Agent) Resume() {
	a.mu.Lock()
	defer a.mu.Unlock()
	if a.State == StatePaused {
		a.State = StateIdle
	}
}
